cli/command/service/logs.go

```diff
@@ -212,43 +235,90 @@ type logWriter struct {
 	opts *logsOptions
 	f    *taskFormatter
 	w    io.Writer
+
+	// detailsCache saves a pre-sorted list of the details keys, so we don't
+	// have to recompute their order every time. this operates under the
+	// assumption that the details keys are always the same (they should be)
+	detailsCache []string
 }
 
 func (lw *logWriter) Write(buf []byte) (int, error) {
-	contextIndex := 0
+	// this works but ONLY because stdcopy calls write a whole line at a time.
+	// if this ends up horribly broken or panics, check to see if stdcopy has
+	// reneged on that asssumption. (@god forgive me)
+	// also this only works because the logs format is, like, barely parsable.
+	// if something changes in the logs format, this is gonna break
+
+	// there should always be at least 2 parts: details and message. if there
+	// is no timestamp, details will be first (index 0) when we split on
+	// spaces. if there is a timestamp, details will be 2nd (`index 1)
+	detailsIndex := 0
 	numParts := 2
 	if lw.opts.timestamps {
-		contextIndex++
+		detailsIndex++
 		numParts++
 	}
 
+	// break up the log line into parts.
 	parts := bytes.SplitN(buf, []byte(" "), numParts)
 	if len(parts) != numParts {
 		return 0, errors.Errorf("invalid context in log message: %v", string(buf))
 	}
-
-	logCtx, err := lw.parseContext(string(parts[contextIndex]))
+	// parse the details out
+	details, err := client.ParseLogDetails(string(parts[detailsIndex]))
+	if err != nil {
+		return 0, err
+	}
+	// and then create a context from the details
+	// this removes the context-specific details from the details map, so we
+	// can more easily print the details later
+	logCtx, err := lw.parseContext(details)
 	if err != nil {
 		return 0, err
 	}
 
 	output := []byte{}
-	for i, part := range parts {
-		// First part doesn't get space separation.
-		if i > 0 {
-			output = append(output, []byte(" ")...)
-		}
-
-		if i == contextIndex {
-			formatted, err := lw.f.format(lw.ctx, logCtx)
-			if err != nil {
-				return 0, err
+	// if we included timestamps, add them to the front
+	if lw.opts.timestamps {
+		output = append(output, parts[0]...)
+		output = append(output, []byte(" ")...)
+	}
+	// add the context, nice and formatted
+	formatted, err := lw.f.format(lw.ctx, logCtx)
+	if err != nil {
+		return 0, err
+	}
+	output = append(output, []byte(fmt.Sprintf("%s    | ", formatted))...)
```

Comment:
The `Sprintf` is being used for simple string concatenation. It could be `formatted + "    | "`.